### ISSUE TITLE
opening_brace generates false positives on closures returning closures

### ISSUE THREAD

**Issue by mekarthedev** [realm/SwiftLint#1728]
When enabled opening_brace inspection, swiftlint generates warning for the following code:

```swift
func map<T, R>(_ t: T, transform: T -> R) -> R {
    return transform(t)
}

func demo() {
    let r = map(42) { t in
        { t + 1 }
    }
    print(r())
}
```

`warning: Opening Brace Spacing Violation: Opening braces should be preceded by a single space and on the same line as the declaration. (opening_brace)` on the line `{ t + 1 }`

The fix expected by opening_brace rule:

```swift
let r = map(42) { t in {
    t + 1
} }
```

Such syntax may cause misunderstanding of what `map(42) { t in` really does. One may overlook that it doesn't immediately adds +1 but does it lazilly. And in overall it doesn't look great.

A temporary fix would be to write `return {`. But it may conflict with `implicit_return` rule.

Another use case:

```swift
perform {
    {
        someLock.lock(); defer { someLock.unlock() }
        // critical section
        // ...
    }()

    // no lock
}
```

Ideally, opening_brace rule should ignore opening braces that are not related to any previous statement or declaration.
